escape_roomba/thread_manager.py

```diff
@@ -0,0 +1,327 @@
+        # Don't react to our own emoji reactions (superfluous).
+        if p.user_id != self._context.client.user.id:
+            await self._async_maybe_update(
+                channel_id=p.channel_id, message_id=p.message_id,
+                emoji=p.emoji)
+
+    async def _on_raw_reaction_clear(self, p):
+        await self._async_maybe_update(
+            channel_id=p.channel_id, message_id=p.message_id)
+
+    async def _on_raw_reaction_clear_emoji(self, payload):
+        await self._async_update_message(
+            channel_id=p.channel_id, message_id=p.message_id, emoji=p.emoji)
+
+    async def _on_guild_channel_delete(self, channel):
+        # Take note when a thread channel is deleted.
+        await self._async_forget_channel(channel)
+
+    async def _on_guild_channel_create(self, channel):
+        # Check if a new channel is a thread channel.
+        await self._async_check_channel(channel)
+        await self._async_fetch_history(channel=channel)
+
+    async def _on_guild_channel_update(self, channel):
+        # Check if a modified channel is a thread channel.
+        await self._async_check_channel(channel)
+
+    #
+    # Internal methods
+    #
+
+    async def _async_maybe_update(self, channel_id=None, channel=None,
+                                  message=None, message_id=None, emoji=None):
+        """Responds to an indication of message change; if the change is
+        relevant, fetches and processes the message's latest version.
+        One each of channel_id/channel and message_id/message must be non-None.
+
+        Args:
+            channel_id: int - ID of channel containing message *OR*
+            channel: discord.TextChannel - channel object containing message
+            message_id: int - ID of message to fetch and process *OR*
+            message: discord.Message - contents of updated message
+            emoji: str, discord.*Emoji - emoji of reaction change (or None)
+        """
+
+        # Handle various ways channel and message can be supplied
+        ci = channel_id or channel.id
+        mi = message_id or message.id
+
+        # Skip irrelevant changes; a change is "relevant" if:
+        # - there is an existing fetch in progress for the message OR
+        # - the message is an existing thread channel origin OR
+        # - the message has a thread emoji reaction OR
+        # - there was a change to thread emoji reactions
+        if ((ci, mi) not in self._update_done and
+            mi not in self._thread_by_origin.get(ci, {}) and
+            (emoji is None or str(emoji) != self._THREAD_EMOJI) and
+            (message is None or
+             all(str(r) != self._THREAD_EMOJI for r in message.reactions))):
+            return
+
+        # To avoid race conditions with asynchronous message fetch requests,
+        # serialize update operations, using a per-message-ID map of Futures.
+        # (Per-message Locks would be easier, but would accumulate over time.)
```

Comment:
It might be helpful to explain here that really you only ever have at most on "pending future" per message (in addition to the one that is being processed by this code), and if one is found, you wait on that to complete before proceeding with the current fetch request.

Reply:
Added some explanation!